Improved Scheduler API

4drian3d opened the pull request:
- Added `Scheduler#buildTask(plugin, Consumer<ScheduledTask>)`
- Added `Scheduler#tasksByPlugin(plugin)`
Allows to obtain the tasks that a plugin has schedule to execute and that are currently active

Resolves #276

4drian3d committed:
Improved Scheduler API

- Added `Scheduler#builder(plugin)`
This method allows a more simplified builder while maintaining the main requirement of the executor plugin
- Added `Scheduler#taskByPlugin(plugin)`
Allows to obtain the tasks that a plugin has sent to execute and that are currently active
- Added `TaskBuilder#task(Consumer<SchuledTask>)`
Allows to specify a task with access to the task itself with the ability to cancel itself

hugmanrique commented on proxy/src/main/java/com/velocitypowered/proxy/scheduler/VelocityScheduler.java:
```diff
@@ -65,12 +68,28 @@ public VelocityScheduler(PluginManager pluginManager) {
+    requireNonNull(plugin, () -> "the plugin cannot be null");
+    checkArgument(pluginManager.fromInstance(plugin).isPresent(), "plugin is not registered");
+    synchronized (tasksByPlugin) {
+      return Set.copyOf(tasksByPlugin.get(plugin));
```
Nit: the `get` call can be moved outside of the synchronized block (https://javadoc.io/doc/com.google.guava/guava/latest/com/google/common/collect/Multimaps.html#synchronizedMultimap(com.google.common.collect.Multimap) (the synchronized block would then contain `Set.copyOf(tasks)`.

hugmanrique commented on proxy/src/test/java/com/velocitypowered/proxy/scheduler/VelocitySchedulerTest.java:
```diff
@@ -65,4 +66,39 @@ void repeatTaskWorks() throws Exception {
+
+    scheduler.builder(FakePluginManager.PLUGIN_A)
+      .task(ScheduledTask::cancel)
+      .delay(100, TimeUnit.MILLISECONDS)
```
This test is time-dependent, and could fail when running with a single core. Could we perform the assertion inside the task itself (and so the task no longer needs to be delayed)?

4drian3d replied:
_With this specific test I doubt very much that it could fail, since I doubt that going from one simple method to another would take 100 milliseconds_

I just understood what you mean with the following comment

hugmanrique commented on proxy/src/test/java/com/velocitypowered/proxy/scheduler/VelocitySchedulerTest.java:
```diff
@@ -65,4 +66,39 @@ void repeatTaskWorks() throws Exception {
+    VelocityScheduler scheduler = new VelocityScheduler(new FakePluginManager());
+
+    ScheduledTask task = scheduler.builder(FakePluginManager.PLUGIN_B)
+        .task(ScheduledTask::cancel)
```
Same as above: we could schedule a task inside the original task that checks its status.

hugmanrique commented on api/src/main/java/com/velocitypowered/api/scheduler/Scheduler.java:
```diff
@@ -23,29 +27,61 @@ public interface Scheduler {
+   * @param plugin the plugin to request the task for
+   * @return the task builder
+   */
+  TaskBuilder builder(@NotNull Object plugin);
```
In the original issue, @astei argued that methods to create a builder that take all mandatory arguments are more idiomatic. Could we use the same approach and replace this by a `buildTask(Object, Consumer<ScheduledTask>)` method?

I rarely see the need for the `TaskBuilder#task` methods below, but I guess it makes the builders reusable (this should be documented). On the other hand, e.g. once a `Runnable` task has been specified, the `Consumer` variant can no longer be called and vice versa. This is a bit inconsistent. An alternative is for the `task` methods to clear the other field (e.g. `TaskBuilder#task(Runnable)` sets the `Consumer` field to `null`). That way we can also remove the precondition in the `schedule` method (it confused me at first).

Somewhat related: I would change the conditions of the form `runnable == null ? use consumer : use runnable` to `runnable != null ? use runnable : use consumer` to prevent bugs in the future.

4drian3d replied:
Now that you mention it, I would have liked the idea that the builder could be reusable, but I find its usage very rare and somewhat confusing

4drian3d committed:
Applied requested changes

- Removed tasks builder method
- Added `Scheduler#buildTask(plugin, Consumer<ScheduledTask>)`

4drian3d committed:
Removed some unused imports

4drian3d committed:
Applied suggested change

4drian3d committed:
Fix possible test bug

astei commented on proxy/src/main/java/com/velocitypowered/proxy/scheduler/VelocityScheduler.java:
```diff
@@ -65,10 +68,28 @@ public VelocityScheduler(PluginManager pluginManager) {
   public TaskBuilder buildTask(Object plugin, Runnable runnable) {
-    checkNotNull(plugin, "plugin");
-    checkNotNull(runnable, "runnable");
+    requireNonNull(runnable, () -> "the scheduled task cannot be null");
```
why change this?

4drian3d replied:
I thought it would be better to use a Supplier to avoid creating an unnecessary string each time this method is called and no argument is null

astei replied:
The string is stored in the constant pool. There is no need to "create an unnecessary string".

astei commented on proxy/src/test/java/com/velocitypowered/proxy/scheduler/VelocitySchedulerTest.java:
```diff
@@ -65,4 +67,36 @@ void repeatTaskWorks() throws Exception {
 
+  @Test
+  void obtainTasksFromPlugin() throws Exception {
+    VelocityScheduler scheduler = new VelocityScheduler(new FakePluginManager());
```
I'm nervous about this test. How do we know it will pass or fail? Would feel more comfortable with a `CountDownLock` and doing an assertion in the test thread.

4drian3d replied:
If you mean a `CountDownLatch`, I just implemented it in the last commit

astei replied:
This is not acceptable. We should NEVER be asserting inside a task.

astei commented on proxy/src/test/java/com/velocitypowered/proxy/scheduler/VelocitySchedulerTest.java:
```diff
@@ -65,4 +67,36 @@ void repeatTaskWorks() throws Exception {
+
+  @Test
+  void testConsumerCancel() throws Exception {
+    VelocityScheduler scheduler = new VelocityScheduler(new FakePluginManager());
```
ditto

astei replied:
This is not acceptable. We should always assert OUTSIDE of any tasks.

astei commented on proxy/src/test/java/com/velocitypowered/proxy/scheduler/VelocitySchedulerTest.java:
```diff
@@ -65,4 +67,36 @@ void repeatTaskWorks() throws Exception {
+  void testConsumerCancel() throws Exception {
+    VelocityScheduler scheduler = new VelocityScheduler(new FakePluginManager());
+
+    scheduler.buildTask(FakePluginManager.PLUGIN_B, actualTask -> {
```
I'd also have to have a condition to check if `actualTask` is the same value as what `TaskBuilder#scheduler()` will return.

4drian3d committed:
Applied more suggested changes

astei commented on proxy/src/test/java/com/velocitypowered/proxy/scheduler/VelocitySchedulerTest.java:
```diff
@@ -65,4 +67,75 @@ void repeatTaskWorks() throws Exception {
+    VelocityScheduler scheduler = new VelocityScheduler(new FakePluginManager());
+    CountDownLatch latch = new CountDownLatch(1);
+
+    TaskContainer consumerTask = new TaskContainer();
```
Why not use `AtomicReference<ScheduledTask>`, which is safer?

4drian3d committed:
Fixed tests inside tasks
